### PR TITLE
Feature/256 lambda support

### PR DESCRIPTION
**Pull request by dorians**
This PR brings support for LAMBDA instruction (https://tezos.gitlab.io/master/whitedoc/michelson.html#xii-full-grammar). This is a very specific instruction because it's the only one which have both expressions and instructions as arguments. In odrer to add support for that feature I needed to join ``MichelsonSimpleInstruction`` and ``MichelsonComplexInstruction`` classses and change the way they are parsed.

### CODE REVIEW

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala [hunk 1]
```diff
@@ -144,21 +149,21 @@ object JsonParser {
...
-        Decoder[JsonStringConstant].widen,
-        Decoder[JsonEmptyExpression].widen
+        Decoder[JsonStringConstant].widen
+      ).reduceLeft(_ or _)
+
+    val decodeInstructionSequence: Decoder[JsonInstructionSequence] = _.as[List[JsonInstruction]].map(JsonInstructionSequence)
+
+    implicit val decodeInstruction: Decoder[JsonInstruction] =
+      List[Decoder[JsonInstruction]](
+        decodeInstructionSequence.widen,
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala]: I think it wouldn't be terrible to inline this?

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala [hunk 2]
```diff
@@ -144,21 +149,21 @@ object JsonParser {
...
-      lazy val isEmptyArray = (_: Json).asArray.exists(_.isEmpty)
-      lazy val isComplexInstruction = (_: HCursor).downField("args").downArray.find(it => isSequence(it.hcursor) || isEmptyArray(it)).succeeded
-
-      if (isSequence(cursor))
-        cursor.as[List[JsonInstruction]].map(JsonInstructionSequence)
-      else if (isComplexInstruction(cursor))
-        cursor.as[JsonComplexInstruction]
-      else
-        cursor.as[JsonSimpleInstruction]
+    implicit def decodeEither[A,B](implicit a: Decoder[A], b: Decoder[B]): Decoder[Either[A,B]] = {
```
Comment (ivanopagano) [inline src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala]: oh my, there's nothing already available in circe to do this? it's surprising

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala [hunk 3]
```diff
@@ -45,9 +44,21 @@ object JsonParser {
...
+   * |                         single type "int"
+   * type "pair" with two arguments
+   *
+   * Empty expression is represented as an empty array in JSON. In code is represented as List[String] as a workaround
+   * because circe doesn't support List[Nothing]
+   *
    * */
-  case class JsonType(prim: String, args: Option[List[JsonExpression]]) extends JsonExpression {
-    override def toMichelsonExpression = MichelsonType(prim, args.getOrElse(List.empty).map(_.toMichelsonExpression))
+  case class JsonType(prim: String, args: Option[List[Either[JsonExpression, List[String]]]]) extends JsonExpression {
```
[the remark this answers is not in the document]
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala]: I found that ``Nil.type`` works, event that ``List[Nothing]`` doesn't :)

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala [hunk 4]
```diff
@@ -89,15 +96,13 @@ object JsonParser {
...
-
-  case class JsonComplexInstruction(prim: String, args: List[List[JsonInstruction]]) extends JsonInstruction {
-    override def toMichelsonInstruction = MichelsonComplexInstruction(prim, args.map {
-      case Nil => MichelsonEmptyInstruction
-      case it => MichelsonInstructionSequence(it.map(_.toMichelsonInstruction))
-    })
+  case class JsonSimpleInstruction(prim: String, args: Option[List[Either[JsonExpression, List[JsonInstruction]]]] = None) extends JsonInstruction {
+    override def toMichelsonInstruction = MichelsonSimpleInstruction(prim, args.getOrElse(List.empty)
+      .map {
+        case Left(it) => it.toMichelsonExpression
```
[the remark this answers is not in the document]
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala]: Yes, I use ``it`` by default for simple expressions. In that case, a simple expression has grown suddenly and I didn't change it. I agree with what you are saying.

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala (hunk 1 above)
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala]: I wanted to keep it consistent with other ones.

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala (hunk 2 above)
Comment (dorians) [reply src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala]: The default encoder looking for ``left`` and ``right`` keys.

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala (hunk 1 above)
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala]: oh I didn't realise

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala (hunk 2 above)
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala]: I see

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala (hunk 3 above)
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala]: awesome

File: src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala (hunk 4 above)
[the remark this answers is not in the document]
Comment (ivanopagano) [reply src/main/scala/tech/cryptonomic/conseil/tezos/michelson/parser/JsonParser.scala]: fine

**Pull request merged**